vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc

```diff
@@ -378,124 +357,145 @@ void LedgerImpl::OnPostData(
   }
 }
 
-std::string LedgerImpl::URIEncode(const std::string& value) {
-  return ledger_client_->URIEncode(value);
-}
-
-void LedgerImpl::GetActivityInfoList(
-    uint32_t start,
-    uint32_t limit,
-    type::ActivityInfoFilterPtr filter,
-    ledger::PublisherInfoListCallback callback) {
-  database()->GetActivityInfoList(
-      start,
-      limit,
-      std::move(filter),
-      callback);
+void LedgerImpl::GetActivityInfoList(uint32_t start,
+                                     uint32_t limit,
+                                     type::ActivityInfoFilterPtr filter,
+                                     PublisherInfoListCallback callback) {
+  WhenReady([this, start, limit, filter = std::move(filter),
+             callback]() mutable {
+    database()->GetActivityInfoList(start, limit, std::move(filter), callback);
+  });
 }
 
-void LedgerImpl::GetExcludedList(ledger::PublisherInfoListCallback callback) {
-  database()->GetExcludedList(callback);
+void LedgerImpl::GetExcludedList(PublisherInfoListCallback callback) {
+  WhenReady([this, callback]() { database()->GetExcludedList(callback); });
 }
 
 void LedgerImpl::SetPublisherMinVisitTime(int duration) {
-  state()->SetPublisherMinVisitTime(duration);
+  WhenReady(
+      [this, duration]() { state()->SetPublisherMinVisitTime(duration); });
 }
 
 void LedgerImpl::SetPublisherMinVisits(int visits) {
-  state()->SetPublisherMinVisits(visits);
+  WhenReady([this, visits]() { state()->SetPublisherMinVisits(visits); });
 }
 
 void LedgerImpl::SetPublisherAllowNonVerified(bool allow) {
-  state()->SetPublisherAllowNonVerified(allow);
+  WhenReady([this, allow]() { state()->SetPublisherAllowNonVerified(allow); });
 }
 
 void LedgerImpl::SetPublisherAllowVideos(bool allow) {
-  state()->SetPublisherAllowVideos(allow);
+  WhenReady([this, allow]() { state()->SetPublisherAllowVideos(allow); });
 }
 
 void LedgerImpl::SetAutoContributionAmount(double amount) {
-  state()->SetAutoContributionAmount(amount);
+  WhenReady([this, amount]() { state()->SetAutoContributionAmount(amount); });
 }
 
 void LedgerImpl::SetAutoContributeEnabled(bool enabled) {
-  state()->SetAutoContributeEnabled(enabled);
+  WhenReady([this, enabled]() { state()->SetAutoContributeEnabled(enabled); });
 }
 
 uint64_t LedgerImpl::GetReconcileStamp() {
+  if (!IsReady())
+    return 0;
```

Comment:
Shouldn't we use `base::Optional<>` instead?

Reply:
Yes, definitely. We should at some point audit the "synchronous" functions on `Ledger` and `LedgerClient` and determine whether they need to be made asynchronous, and if not whether they need to return an `optional` (I think we're supposed to use `absl::optional<T>` now).